FlexibleDiff/SectionedChangeset.swift

```diff
@@ -82,57 +82,63 @@ public struct SectionedChangeset {
 	) {
 		let metadata = Changeset(previous: previous, current: current, identifier: sectionIdentifier, areEqual: areMetadataEqual)
 
-		let moveSourceLookup = Dictionary(uniqueKeysWithValues: metadata.moves.lazy.map { ($0.destination, $0.source) })
+		let moveDestinationLookup = Dictionary(uniqueKeysWithValues: metadata.moves.lazy.map { ($0.source, $0.destination) })
 		let mutatedMoveDests = Set(metadata.moves.lazy.filter { $0.isMutated }.map { $0.destination })
 
-		let allInsertions = metadata.inserts.union(IndexSet(moveSourceLookup.keys))
-		let allRemovals = metadata.removals.union(IndexSet(moveSourceLookup.values))
+		let allInsertions = metadata.inserts.union(IndexSet(moveDestinationLookup.values))
+		let allRemovals = metadata.removals.union(IndexSet(moveDestinationLookup.keys))
 
 		mutatedSections = Array()
 		mutatedSections.reserveCapacity(Int(current.count))
 
 		var moves: [Changeset.Move] = []
 		var mutations = IndexSet()
 
-		for (offset, section) in current.enumerated() where !metadata.inserts.contains(offset) {
-			let predeletionOffset: Int
+		for (sourceOffset, section) in previous.enumerated() where !metadata.removals.contains(sourceOffset) {
+			let destinationOffset: Int
 			let isMove: Bool
 
-			if let moveSource = moveSourceLookup[offset] {
-				predeletionOffset = moveSource
+			if let moveDestination = moveDestinationLookup[sourceOffset] {
+				destinationOffset = moveDestination
 				isMove = true
 			} else {
-				let preinsertionOffset = offset - allInsertions.count(in: 0 ..< offset)
-				predeletionOffset = preinsertionOffset + allRemovals.count(in: 0 ... preinsertionOffset)
+				let postRemovalOffset = sourceOffset - allRemovals.count(in: 0 ..< sourceOffset)
+
+				var insertionsBeforeOffset = IndexSet()
+				for insertion in allInsertions where insertion <= postRemovalOffset + insertionsBeforeOffset.count {
+					insertionsBeforeOffset.insert(insertion)
```

Comment:
Could you break the loop when the predicate is not satisfied? There is little value to continue iterating afterwards.